### ISSUE TITLE
Data loss after crash

### ISSUE THREAD

**Issue by ghost** [skyjake/lagrange#24]
History, bookmarks, identities, and open tabs are lost after a crash. Maybe it would be a good idea to flush these out to disk periodically? Lost an identity that I needed to authenticate.

Can't currently replicate the crash. Found this in the terminal, unsure if related to the crash:
```
140065896150784:error:1408F119:SSL routines:ssl3_get_record:decryption failed or bad record mac:../ssl/record/ssl3_record.c:676:
140065334228736:error:1408F119:SSL routines:ssl3_get_record:decryption failed or bad record mac:../ssl/record/ssl3_record.c:676:
```

**Comment by skyjake** [comment]
When it comes to identities, they are immediately written to files after creation so those should be recoverable? In fact, Lagrange should notice the files are in the `idents/` subdirectory and import them automatically. (Of course, this only applies to non-temporary identities.)

Bookmarks, yes, those should definitely be saved as soon as changes are made via the UI. 

Open tabs and history are currently only saved at shutdown because it may involve a (relatively) large data dump to disk. I'll have to review the performance implications and see if using a background thread is appropriate. In any case, yeah they should be saved periodically.

> Found this in the terminal, unsure if related to the crash:

If you notice a specific URL where this happens, let me know.

**Comment by ghost** [comment]
The current behavior for open tabs and history are understandable, the implications of doing it any other way seem complex.

Regarding the identity, I remember setting "temporary" to YES, could that explain why it's gone after the crash?

The bookmarks are definitely gone, however.

Desperately looking for the URL that made it crash...

**Comment by skyjake** [comment]
> Regarding the identity, I remember setting "temporary" to YES, could that explain why it's gone after the crash?

Yes, that would do it. It's poorly documented currently, but temporary identities are kept in runtime memory only and never saved to disk. I'll add a note in the UI making this clear as it's pretty important.

**Comment by winduptoy** [comment]
I can now confirm that bookmarks are not actually flushed until exiting.

**Comment by skyjake** [comment]
That is correct, yes. However, I just pushed a fix for it (see above). 🙂

**Comment by skyjake** [comment]
> The current behavior for open tabs and history are understandable, the implications of doing it any other way seem complex.

Now that I think about it, it would be pretty easy to periodically write the URLs only and do a full cache dump only at shutdown. That way we can retain the full navigation stack even after crashes, and don't have to worry about writing large files to disk.

**Comment by skyjake** [comment]
> History, bookmarks, identities, and open tabs are lost after a crash. 

Status:

- History: ✅ fixed in 64b9848f878ccd75aae75bf4d8d3f8a364f28ddc
- Bookmarks: ✅ fixed, now saved after changes 
- Identities: ✅ working as intended, added a note in the UI about temporary identities
- Tabs: not fixed

**Comment by ghost** [comment]
:+1:

I can't duplicate the crash. If it crops up again I'll make a new issue.
